File: src/cadforge/tools/registry.py
"""Tool registry for CadForge.

Defines all available tools with their JSON schemas for the Anthropic API.
"""

from __future__ import annotations

from typing import Any


def get_tool_definitions() -> list[dict[str, Any]]:
    """Return all tool definitions in Anthropic API tool format."""
    return [
        {
            "name": "ExecuteCadQuery",
            "description": (
                "Execute CadQuery Python code to create 3D models. "
                "The code runs in a sandboxed environment with cadquery (cq), math, and numpy (np) available. "
                "Assign the final workpiece to 'result' to export it. "
                "Returns mesh statistics and the output file path."
            ),
            "input_schema": {
                "type": "object",
                "properties": {
                    "code": {
                        "type": "string",
                        "description": "CadQuery Python code to execute",
                    },
                    "output_name": {
                        "type": "string",
                        "description": "Name for the output file (without extension)",
                        "default": "model",
                    },
                    "format": {
                        "type": "string",
                        "enum": ["stl", "step"],
                        "description": "Export format",
                        "default": "stl",
                    },
                },
                "required": ["code"],
            },
        },
        {
            "name": "ReadFile",
            "description": "Read the contents of a file.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "path": {
                        "type": "string",
                        "description": "Path to the file to read",
                    },
                },
                "required": ["path"],
            },
        },
        {
            "name": "WriteFile",
            "description": "Write content to a file, creating directories as needed.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "path": {
                        "type": "string",
                        "description": "Path to write the file",
                    },
                    "content": {
                        "type": "string",
                        "description": "Content to write",
                    },
                },
                "required": ["path", "content"],
            },
        },
        {
            "name": "ListFiles",
            "description": "List files and directories matching a glob pattern.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "pattern": {
                        "type": "string",
                        "description": "Glob pattern (e.g., '**/*.py', 'vault/*.md')",
                        "default": "*",
                    },
                    "path": {
                        "type": "string",
                        "description": "Base directory to search from",
                        "default": ".",
                    },
                },
                "required": [],
            },
        },
        {
            "name": "SearchVault",
            "description": (
                "Search the knowledge vault using hybrid vector + full-text search. "
                "Returns relevant sections from materials, design rules, CadQuery patterns, "
                "and printer profiles."
            ),
            "input_schema": {
                "type": "object",
                "properties": {
                    "query": {
                        "type": "string",
                        "description": "Natural language search query",
                    },
                    "tags": {
                        "type": "array",
                        "items": {"type": "string"},
                        "description": "Optional tags to filter results (e.g., ['PLA', 'material'])",
                    },
                    "limit": {
                        "type": "integer",
                        "description": "Maximum number of results",
                        "default": 5,
                    },
                },
                "required": ["query"],
            },
        },
        {
            "name": "AnalyzeMesh",
            "description": (
                "Analyze a 3D mesh file (STL/3MF) for watertightness, volume, "
                "surface area, bounding box, and potential issues."
            ),
            "input_schema": {
                "type": "object",
                "properties": {
                    "path": {
                        "type": "string",
                        "description": "Path to the mesh file",
                    },
                },
                "required": ["path"],
            },
        },
        {
            "name": "ShowPreview",
            "description": "Open an interactive 3D viewer for an STL file.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "path": {
                        "type": "string",
                        "description": "Path to the STL file to preview",
                    },
                },
                "required": ["path"],
            },
        },
        {
            "name": "ExportModel",
            "description": "Export a previously generated model to STEP or 3MF format.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "source": {
                        "type": "string",
                        "description": "Path to the source STL or model file",
                    },
                    "format": {
                        "type": "string",
                        "enum": ["step", "3mf"],
                        "description": "Target export format",
                    },
                    "name": {
                        "type": "string",
                        "description": "Output filename (without extension)",
                    },
                },
                "required": ["source", "format"],
            },
        },
        {
            "name": "Bash",
            "description": (
                "Execute a shell command. Use for git operations, file system tasks, "
                "and other terminal operations."
            ),
            "input_schema": {
                "type": "object",
                "properties": {
                    "command": {
                        "type": "string",
                        "description": "The shell command to execute",
                    },
                    "timeout": {
                        "type": "integer",
                        "description": "Timeout in seconds",
                        "default": 120,
                    },
                },
                "required": ["command"],
            },
        },
        {
            "name": "GetPrinter",
            "description": "Get the active printer profile with build constraints.",
            "input_schema": {
                "type": "object",
                "properties": {},
                "required": [],
            },
        },
        {
            "name": "SearchWeb",
            "description": "Search the web for engineering references and documentation.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "query": {
                        "type": "string",
                        "description": "Search query",
                    },
                },
                "required": ["query"],
            },
        },
    ]


def get_tool_names() -> list[str]:
    """Get list of all tool names."""
    return [t["name"] for t in get_tool_definitions()]


def get_tool_definition(name: str) -> dict[str, Any] | None:
    """Get a specific tool definition by name."""
    for tool in get_tool_definitions():
        if tool["name"] == name:
            return tool
    return None

File: packages/cli/src/tools/registry.ts
/**
 * Tool definitions — Anthropic-format JSON schemas for all tools.
 */

export interface ToolDefinition {
  name: string;
  description: string;
  input_schema: Record<string, unknown>;
}

/**
 * Where a tool executes.
 */
export type ToolLocation = 'local' | 'remote';

export interface ToolRegistryEntry {
  definition: ToolDefinition;
  location: ToolLocation;
}

/**
 * All registered tools.
 */
export const TOOL_REGISTRY: ToolRegistryEntry[] = [
  {
    definition: {
      name: 'ReadFile',
      description: 'Read the contents of a file',
      input_schema: {
        type: 'object',
        properties: {
          path: { type: 'string', description: 'File path to read' },
        },
        required: ['path'],
      },
    },
    location: 'local',
  },
  {
    definition: {
      name: 'WriteFile',
      description: 'Write content to a file',
      input_schema: {
        type: 'object',
        properties: {
          path: { type: 'string', description: 'File path to write' },
          content: { type: 'string', description: 'Content to write' },
        },
        required: ['path', 'content'],
      },
    },
    location: 'local',
  },
  {
    definition: {
      name: 'ListFiles',
      description: 'List files matching a glob pattern',
      input_schema: {
        type: 'object',
        properties: {
          pattern: { type: 'string', description: 'Glob pattern' },
          path: { type: 'string', description: 'Base directory' },
        },
        required: ['pattern'],
      },
    },
    location: 'local',
  },
  {
    definition: {
      name: 'Bash',
      description: 'Execute a shell command',
      input_schema: {
        type: 'object',
        properties: {
          command: { type: 'string', description: 'Shell command to execute' },
          timeout: { type: 'number', description: 'Timeout in milliseconds' },
        },
        required: ['command'],
      },
    },
    location: 'local',
  },
  {
    definition: {
      name: 'ExecuteCadQuery',
      description: 'Execute CadQuery code to create or modify 3D models',
      input_schema: {
        type: 'object',
        properties: {
          code: { type: 'string', description: 'CadQuery Python code' },
          output_name: { type: 'string', description: 'Output filename (without extension)' },
          format: { type: 'string', enum: ['stl', 'step'], description: 'Export format' },
        },
        required: ['code'],
      },
    },
    location: 'remote',
  },
  {
    definition: {
      name: 'AnalyzeMesh',
      description: 'Analyze a mesh file for geometry metrics and manufacturing issues',
      input_schema: {
        type: 'object',
        properties: {
          path: { type: 'string', description: 'Path to mesh file' },
        },
        required: ['path'],
      },
    },
    location: 'remote',
  },
  {
    definition: {
      name: 'SearchVault',
      description: 'Search the knowledge vault for relevant information',
      input_schema: {
        type: 'object',
        properties: {
          query: { type: 'string', description: 'Search query' },
          tags: { type: 'array', items: { type: 'string' }, description: 'Filter by tags' },
          limit: { type: 'number', description: 'Max results' },
        },
        required: ['query'],
      },
    },
    location: 'remote',
  },
  {
    definition: {
      name: 'ShowPreview',
      description: 'Open 3D preview of a mesh file',
      input_schema: {
        type: 'object',
        properties: {
          path: { type: 'string', description: 'Path to STL file' },
        },
        required: ['path'],
      },
    },
    location: 'remote',
  },
  {
    definition: {
      name: 'ExportModel',
      description: 'Export a model to a different format',
      input_schema: {
        type: 'object',
        properties: {
          source: { type: 'string', description: 'Source model path' },
          format: { type: 'string', enum: ['stl', 'step', '3mf'], description: 'Target format' },
          name: { type: 'string', description: 'Output filename' },
        },
        required: ['source'],
      },
    },
    location: 'remote',
  },
  {
    definition: {
      name: 'RenderModel',
      description: 'Render a 3D model (STL file) to PNG images from multiple camera angles',
      input_schema: {
        type: 'object',
        properties: {
          stl_path: { type: 'string', description: 'Path to the STL file to render' },
          output_dir: { type: 'string', description: 'Output directory for PNGs (defaults to same as STL)' },
          include_base64: { type: 'boolean', description: 'Include base64 image data in response' },
        },
        required: ['stl_path'],
      },
    },
    location: 'remote',
  },
  {
    definition: {
      name: 'GetPrinter',
      description: 'Get the active printer profile',
      input_schema: {
        type: 'object',
        properties: {},
      },
    },
    location: 'local',
  },
  {
    definition: {
      name: 'Task',
      description: 'Delegate a task to a subagent',
      input_schema: {
        type: 'object',
        properties: {
          agent_type: { type: 'string', enum: ['explore', 'plan', 'cad', 'design', 'competitive'], description: 'Subagent type' },
          prompt: { type: 'string', description: 'Task prompt' },
          context: { type: 'string', description: 'Additional context' },
        },
        required: ['agent_type', 'prompt'],
      },
    },
    location: 'local',
  },
];

/**
 * Get tool definitions in Anthropic API format.
 */
export function getToolDefinitions(): ToolDefinition[] {
  return TOOL_REGISTRY.map((entry) => entry.definition);
}
